src/chainstate/stacks/index/cache.rs

```diff
@@ -0,0 +1,877 @@
+use std::iter::FromIterator;
+use std::marker::PhantomData;
+use std::ops::{Deref, DerefMut};
+use std::os;
+use std::path::{Path, PathBuf};
+use std::time::SystemTime;
+use std::{cmp, error};
+
+use rusqlite::{
+    types::{FromSql, ToSql},
+    Connection, Error as SqliteError, ErrorCode as SqliteErrorCode, OpenFlags, OptionalExtension,
+    Transaction, NO_PARAMS,
+};
+
+use chainstate::stacks::index::bits::{
+    get_node_byte_len, get_node_hash, read_block_identifier, read_hash_bytes, read_node_hash_bytes,
+    read_nodetype, read_root_hash, write_nodetype_bytes,
+};
+use chainstate::stacks::index::node::{
+    clear_backptr, is_backptr, set_backptr, TrieNode, TrieNode16, TrieNode256, TrieNode4,
+    TrieNode48, TrieNodeID, TrieNodeType, TriePath, TriePtr,
+};
+use chainstate::stacks::index::Error;
+use chainstate::stacks::index::TrieLeaf;
+use chainstate::stacks::index::{trie_sql, ClarityMarfTrieId, MarfTrieId};
+use util_lib::db::sql_pragma;
+use util_lib::db::sqlite_open;
+use util_lib::db::tx_begin_immediate;
+use util_lib::db::tx_busy_handler;
+use util_lib::db::Error as db_error;
+use util_lib::db::SQLITE_MMAP_SIZE;
+
+use stacks_common::types::chainstate::BlockHeaderHash;
+use stacks_common::types::chainstate::BLOCK_HEADER_HASH_ENCODED_SIZE;
+use stacks_common::types::chainstate::{TrieHash, TRIEHASH_ENCODED_SIZE};
+
+/// Fully-qualified address of a Trie node.  Includes both the block ID and the pointer within the
+/// block's blob as to where it is stored.
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub struct TrieNodeAddr(u32, TriePtr);
+
+/// Cache state for all node caching strategies.
+pub struct TrieCacheState<T: MarfTrieId> {
+    /// Mapping between trie blob IDs (i.e. rowids) and the MarfTrieId of the trie.  Contents are
+    /// never evicted, since the size of this map grows only at the rate of new Stacks blocks.
+    block_hash_cache: HashMap<u32, T>,
+
+    /// Mapping between trie blob hashes and their IDs
+    block_id_cache: HashMap<T, u32>,
+
+    /// cached nodes
+    node_cache: HashMap<TrieNodeAddr, TrieNodeType>,
+    /// cached trie root hashes
+    hash_cache: HashMap<TrieNodeAddr, TrieHash>,
+}
+
+impl<T: MarfTrieId> TrieCacheState<T> {
+    pub fn new() -> TrieCacheState<T> {
+        TrieCacheState {
+            block_hash_cache: HashMap::new(),
+            block_id_cache: HashMap::new(),
+            node_cache: HashMap::new(),
+            hash_cache: HashMap::new(),
+        }
+    }
+
+    /// Obtain a possibly-cached node and its hash.
+    /// Only return data if we have *both* the node and hash
+    pub fn load_node_and_hash(
+        &self,
+        block_id: u32,
+        trieptr: &TriePtr,
+    ) -> Option<(TrieNodeType, TrieHash)> {
+        match (
+            self.load_node(block_id, trieptr),
+            self.load_node_hash(block_id, trieptr),
+        ) {
+            (Some(ref node), Some(ref hash)) => Some(((*node).clone(), (*hash).clone())),
```

Comment:
This is a double clone, right? `self.load_node()` and `self.load_node_hash()` do one clone, and this method does another clone.

You can avoid double clones by returning references, which would also let callers that don't need ownership to avoid a clone altogether.

Reply:
Why yes it is.  Will fix.

The implementation always needs an owned copy of the node when it's loaded, which is why there's no option to return a reference to it.  There is a method for getting the reference to a hash, however, since it's not always the case that the caller needs an owned copy.